### DISCUSSION TITLE
Pass in the Entire Context to the Validator

### DISCUSSION THREAD

**Discussion by abhirathmahipal** [proteanhq/protean#371] [Ideas]
**Current API**

I've left some annotations in the code below, please read the comments as well :)
```python

class ScoreValueValidator:
    """Validate the range of score value to be between 0.5 and 5"""

    
    def __call__(self, value):
        # value only contains the current field's value. 
        # How do we write validations which might require other fields as well?

        if value and not all(0 <= value <= 5 for value in value.values()):
            raise ValidationError("is invalid")


@domain.aggregate
class Assignment:
    competition_id = Identifier(required=True)
    group_id = Identifier(required=True)
    judge_id = Identifier(required=True)
    entry_id = Identifier(required=True)
    score = Dict(validators=[ScoreValueValidator()])
```

**Suggested API**

```python

class ScoreValueValidator:
    """Validate the range of score value to be between 0.5 and 5"""

  
    def __call__(self, value, obj):
       # demonstrating validation using other fields
       if not obj.group_id:
            return

        if value and not all(0 <= value <= 5 for value in value.values()):
            raise ValidationError("is invalid")
```

The same should probably apply for `Aggregates, Entities, Value Objects and Commands` in my opinion.

**Comment by subhashb** [comment]
Since such a validation involves multiple fields, it is best placed at the model level. The `object.clean()` method is a good place for model-level field validations, along with in-built and custom field validators. Django adopts a similar validation [technique](https://docs.djangoproject.com/en/dev/ref/models/instances/#validating-objects-1).

As a contra, supporting access to other fields within validators will result in more complex code because:
* The validation is buried within a validator, so is easily missed when understangind the model and its invariants
* The element's structure is tightly coupled with the validator restricting its reusability

**Reply by abhirathmahipal** [reply]
@sagaragarwal94

**Comment by subhashb** [comment]
Hi @abhirathmahipal,

Following up on this a while later. My earlier point still holds: validation that spans fields belongs at the model level. A single field's validator is the wrong place for it, since the rule gets buried in one field and the field's structure ends up coupled to it.

What I called `object.clean()` back then is now expressed with invariants. You declare `@invariant.post` (or `@invariant.pre`) methods on the aggregate, entity, or value object, and they run against the fully-formed instance, so they can read every field and enforce cross-field rules. That gives you the case your suggested `__call__(self, value, obj)` was reaching for, without pushing other-field access into a single validator.

For your example, the range check on the `score` dict can stay a field validator, and any rule that also depends on `group_id` or the other fields moves into an `@invariant.post` method on `Assignment`.

The details are in the invariants guide: https://docs.proteanhq.com/guides/domain-behavior/invariants/

Let me know if this does not cover your need.
